src/renderer/canvas/label.ts

```diff
@@ -58,40 +72,160 @@ export class Label {
   }
 }
 
-export const drawLabel = (context: CanvasRenderingContext2D, label: Label) => {
+export const drawLabel = (context: CanvasRenderingContext2D, label: Label, type: 'node' | 'edge') => {
   const isDrawable = label.textLines.length > 0 && label.fontSize > 0;
   if (!isDrawable || !label.position) {
     return;
   }
 
-  drawTextBackground(context, label);
-  drawText(context, label);
+  const borderPoints: BorderPoint[] | undefined = drawTextBackgroundAndBorder(context, label, type);
+
+  if (!borderPoints) {
+    return;
+  }
+
+  drawTextBackground(context, borderPoints, label);
+  drawTextBackgroundBorder(context, borderPoints, label);
+  drawText(context, borderPoints, label, type);
 };
 
-const drawTextBackground = (context: CanvasRenderingContext2D, label: Label) => {
-  if (!label.properties.fontBackgroundColor || !label.position) {
+const drawTextBackgroundAndBorder = (
```

Comment:
The name of this function and what it does doesn't match. It tells that it is doing the drawing, but it is actually getting the points for the background of the text.
When I get into it much deeper, those are actually bounding boxes rather than border points. But this is nitpicking.